core/vm/evm.go

```diff
@@ -202,24 +196,34 @@ func (self *EVM) Main(trx *Transaction) (ret ExecutionResult) {
 		gas_cap = uint64(math.MaxUint64)
 	} else {
 		if !BalanceGTE(caller, gas_fee) {
-			ret.ConsensusErr = ErrInsufficientBalanceForGas
-			return
+			caller_balance := caller.GetBalance()
+			availiable_funds_gas := bigutil.Div(caller_balance, gas_price)
+			caller.SubBalance(bigutil.Mul(availiable_funds_gas, gas_price))
+
+			return consensusErr(availiable_funds_gas.Uint64(), ErrInsufficientBalanceForGas)
 		}
+		caller.SubBalance(gas_fee)
+	}
+
+	// Check if tx.nonce <= sender_nonce
+	if self.trx.Nonce.Cmp(sender_nonce) <= 0 {
+		return consensusErr(gas_cap, ErrNonceTooLow)
 	}
 
+	// Nonce skipping is permanently enabled now. Uncomment this part to have strict nonce ordering
+	// Check if tx.nonce > sender_nonce + 1
+	// if self.trx.Nonce.Cmp(bigutil.Add(sender_nonce, big.NewInt(1))) > 0 {
+	// 	ret.ConsensusErr = ErrNonceTooHigh
+	// 	return
+	// }
+
 	gas_intrinsic, err := IntrinsicGas(self.trx.Input, contract_creation, self.rules.IsHomestead)
 	if err != nil {
-		ret.ConsensusErr = util.ErrorString(err.Error())
-		return
+		return consensusErr(gas_cap, util.ErrorString(err.Error()))
 	}
 
 	if gas_cap < gas_intrinsic {
-		ret.ConsensusErr = ErrIntrinsicGas
-		return
-	}
-
-	if self.trx.From != common.ZeroAddress {
-		caller.SubBalance(gas_fee)
```

Comment:
We could use zero_address here only for call and estimation. And all changes would be anyway reverted, so I don't think we need to cover it here

Reply:
And actually this SubBalance aren't called for zero_address because they are in this else section 
````	
if self.trx.From == common.ZeroAddress {
	gas_cap = uint64(math.MaxUint64)
} else {
	if !BalanceGTE(caller, gas_fee) {
		caller_balance := caller.GetBalance()
		availiable_funds_gas := bigutil.Div(caller_balance, gas_price)
		caller.SubBalance(bigutil.Mul(availiable_funds_gas, gas_price))

		return consensusErr(availiable_funds_gas.Uint64(), ErrInsufficientBalanceForGas)
	}
	caller.SubBalance(gas_fee)
}
```